vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -21,370 +22,414 @@
+                            max_fill = std::max(max_fill, distance);
+                        } else {
+                            couldnt_fill = true;
+                        }
+                    }
+                }
+#if !defined(FILL_LIMIT)
+                if (couldnt_fill) {
+                    break;
+                }
+#endif
             }
-        }
-
-        return UNDEFINED;
-    }
 
-  private:
-    int start_x_;
-    int start_y_;
-    t_rr_type rr_type_;
-    int seg_index_;
-    size_t index_;
-};
+            if (!couldnt_fill) {
+                VTR_LOG("At %d -> %d: max_fill = %d, delay_penalty = %e\n", seg, box, max_fill, delay_penalty);
+            }
 
-// Minimum size of search for channels to profile.  kMinProfile results
-// in searching x = [0, kMinProfile], and y = [0, kMinProfile[.
-//
-// Making this value larger will increase the sample size, but also the runtime
-// to produce the lookahead.
-static constexpr int kMinProfile = 1;
+            // write back the missing entries
+            for (auto& xy_entry : missing) {
+                matrix[std::get<0>(xy_entry)][std::get<1>(xy_entry)] = std::get<2>(xy_entry);
+            }
 
-// Maximum size of search for channels to profile.  Once search is outside of
-// kMinProfile distance, lookahead will stop searching once:
-//  - At least one channel has been profiled
-//  - kMaxProfile is exceeded.
-static constexpr int kMaxProfile = 7;
+            if (couldnt_fill) {
+                VTR_LOG_WARN("Couldn't fill holes in the cost matrix for %d -> %ld, %d x %d bounding box\n",
+                             seg, box, seg_box_bounds.width(), seg_box_bounds.height());
+#if !defined(FILL_LIMIT)
+                for (unsigned y = 0; y < matrix.dim_size(1); y++) {
+                    for (unsigned x = 0; x < matrix.dim_size(0); x++) {
+                        VTR_ASSERT(!matrix[x][y].valid());
+                    }
+                }
+#endif
+            }
+        }
+    }
+}
 
-static int search_at(int iseg, int start_x, int start_y, t_routing_cost_map* cost_map) {
+// prints an ASCII diagram of each cost map for a segment type (debug)
+// o => above average
+// . => at or below average
+// * => invalid (missing)
+void CostMap::print(int iseg) const {
     const auto& device_ctx = g_vpr_ctx.device();
-
-    int count = 0;
-    int dx = 0;
-    int dy = 0;
-
-    while ((count == 0 && dx < kMaxProfile) || dy <= kMinProfile) {
-        if (start_x + dx >= device_ctx.grid.width()) {
-            break;
-        }
-        if (start_y + dy >= device_ctx.grid.height()) {
-            break;
+    for (size_t box_id = 0;
+         box_id < device_ctx.connection_boxes.num_connection_box_types();
+         box_id++) {
+        auto& matrix = cost_map_[iseg][box_id];
+        if (matrix.dim_size(0) == 0 || matrix.dim_size(1) == 0) {
+            printf("cost EMPTY for box_id = %lu\n", box_id);
```

Comment:
Use VTR_LOG rather than printf so that VPR's logging is consistent.  Right now, this prints to stdout, but not to the VPR log file (which defaults to vpr_stdout.log).

Reply:
This is only for debugging, and VTR_LOG would mangle the output because it's not line-by-line.

Reply:
> because it's not line-by-line.

That's not true?

Reply:
Right. The warnings add stuff, but VTR_LOG doesn't seem to. Fixed.